runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionStateTrackerTest.java

```diff
@@ -123,12 +158,115 @@ private void assertElementProcessingTimeCounter(NameContext step, int millis, in
+    // Adding test for the full thread dump, but since we can't mock
+    // Thread.getAllStackTraces(), we are starting a background thread
+    // to verify the full thread dump.
+    Thread backgroundThread =
+        new Thread("backgroundThread") {
+          @Override
+          public void run() {
+            try {
+              Thread.sleep(Long.MAX_VALUE);
+            } catch (InterruptedException e) {
+              // exiting the thread
+            }
+          }
+        };
+
+    backgroundThread.start();
+    try {
+      // Full thread dump should be performed, because we never performed
+      // a full thread dump before, and the lull duration is more than 20
+      // minutes.
+      tracker.reportBundleLull(mockThread, 30 * 60 * 1000);
+      verifyLullLog();
+
+      // Full thread dump should not be performed because the last dump
+      // was only 5 minutes ago.
+      clock.setTime(clock.currentTimeMillis() + Duration.standardMinutes(5L).getMillis());
+      tracker.reportBundleLull(mockThread, 30 * 60 * 1000);
+      verifyLullLog();
+
+      // Full thread dump should not be performed because the lull duration
+      // is only 6 minutes.
+      clock.setTime(clock.currentTimeMillis() + Duration.standardMinutes(16L).getMillis());
+      tracker.reportBundleLull(mockThread, 6 * 60 * 1000);
+      verifyLullLog();
+
+      // Full thread dump should be performed, because it has been 21 minutes
+      // since the last dump, and the lull duration is more than 20 minutes.
+      clock.setTime(clock.currentTimeMillis() + Duration.standardMinutes(16L).getMillis());
+      tracker.reportBundleLull(mockThread, 30 * 60 * 1000);
+      verifyLullLog();
+    } finally {
+      // Cleaning up the background thread.
+      backgroundThread.interrupt();
+      backgroundThread.join();
+    }
+  }
+
+  private void verifyLullLog() throws IOException {
+    File[] files = logFolder.listFiles();
+    assertThat(files, Matchers.arrayWithSize(1));
+    File logFile = files[0];
+    List<String> lines = Files.readAllLines(logFile.toPath());
+
+    String warnLines =
+        Joiner.on("\n").join(Iterables.filter(lines, line -> line.contains("\"WARN\"")));
+    assertThat(
+        warnLines,
+        Matchers.allOf(
+            Matchers.containsString("Operation ongoing in bundle for at least"),
+            Matchers.containsString(" without completing"),
+            Matchers.containsString("Processing times in each step"),
+            Matchers.containsString(
+                "org.apache.beam.runners.dataflow.worker.DataflowExecutionContext$DataflowExecutionStateTracker"),
+            Matchers.containsString("userpackage.SomeUserDoFn.helperMethod"),
+            Matchers.not(Matchers.containsString(SimpleDoFnRunner.class.getName()))));
+
+    String infoLines =
+        Joiner.on("\n").join(Iterables.filter(lines, line -> line.contains("\"INFO\"")));
+    assertThat(
+        infoLines,
+        Matchers.not(
```

Comment:
it seems from this we are always verifying that a full thread dump is not performed.  The comments in the tests sometimes say we want a full thread dump although looking at code I don't see when we'd ever get one.

Reply:
Fixed the comments.